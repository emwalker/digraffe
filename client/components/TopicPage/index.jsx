import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

type Props = {
  topic: {
    name: string,
  }
}

const TopicPage = ({ topic }: Props) => (
  <div>
    <h1>Topic: {topic ? topic.name : 'Nemo'}</h1>
  </div>
)

export const query = graphql`
query TopicPage_query_Query(
  $orgResourceId: String!,
  $topicResourceId: String!
) {
  organization(resourceId: $orgResourceId) {
    topic(resourceId: $topicResourceId) {
      ...TopicPage_topic
    }
  }
}`

export default createFragmentContainer(TopicPage, graphql`
  fragment TopicPage_topic on Topic {
    name
  }
`)
