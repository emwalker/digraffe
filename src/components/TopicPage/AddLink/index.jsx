// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import createLinkMutation from '../../../mutations/createLinkMutation'
import type { RelayProps } from '../../types'

/* eslint jsx-a11y/label-has-for: 0 */

type State = {
  url: string,
}

class AddLink extends Component<RelayProps, State> {
  state = {
    url: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createLink()
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const { resourceId: organizationId } = this.props.organization

    createLinkMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        organizationId,
        url: this.state.url,
        topicIds: [this.props.topic.resourceId],
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <div>
      <dl className="form-group">
        <dt>
          <label htmlFor="create-link-url">Add link</label>
        </dt>
        <dd>
          <input
            className="form-control test-link-url input-sm"
            id="create-link-url"
            onChange={this.updateUrl}
            onKeyPress={this.onKeyPress}
            placeholder="Url"
            value={this.state.url}
          />
        </dd>
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_organization on Organization {
    resourceId
  }

  fragment AddLink_topic on Topic {
    id
    resourceId
  }
`)
