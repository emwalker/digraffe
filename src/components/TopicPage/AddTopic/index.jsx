// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import upsertTopicMutation from 'mutations/upsertTopicMutation'
import type { RelayProps } from '../../types'

/* eslint jsx-a11y/label-has-for: 0 */

type State = {
  name: string,
}

type Props = RelayProps & {
  viewer: {
    defaultRepository: {
      id: ID,
    },
  },
}

class AddTopic extends Component<Props, State> {
  state = {
    name: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createTopic()
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_childTopics',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'topicEdge',
    }]
  }

  get repositoryId(): ?string {
    if (this.props.viewer)
      return this.props.viewer.defaultRepository.id
    return null
  }

  updateName = (event: Object) => {
    this.setState({ name: event.currentTarget.value })
  }

  createTopic() {
    upsertTopicMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        name: this.state.name,
        repositoryId: this.repositoryId,
        topicIds: [this.props.topic.id],
      },
    )
    this.setState({ name: '' })
  }

  render = () => (
    <div>
      <dl className="form-group">
        <dt>
          <label htmlFor="create-topic-name">Add subtopic</label>
        </dt>
        <dd>
          <input
            className="form-control test-topic-name input-sm"
            id="create-topic-name"
            onChange={this.updateName}
            onKeyPress={this.onKeyPress}
            placeholder="Name or description"
            value={this.state.name}
          />
        </dd>
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddTopic, graphql`
  fragment AddTopic_topic on Topic {
    id
  }
`)
