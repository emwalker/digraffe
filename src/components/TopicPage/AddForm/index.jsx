// @flow
import React, { Component, Fragment } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import classNames from 'classnames'

import type { Relay, TopicType, UserType } from 'components/types'
import AddTopic from './AddTopic'
import AddLink from './AddLink'
import SelectRepository from './SelectRepository'
import './index.css'

type Props = {
  relay: Relay,
  topic: TopicType,
  viewer: UserType,
}

class AddForm extends Component<Props> {
  get className(): string {
    return classNames(
      'border',
      'rounded-1',
      'px-md-2',
      'px-3',
      'mt-3',
      { 'private-repo': this.isPrivateRepo },
    )
  }

  get isPrivateRepo(): boolean {
    const repo = this.props.viewer.selectedRepository
    if (!repo) return false
    return repo.isPrivate
  }

  get repoSelected(): boolean {
    return !!this.props.viewer.selectedRepository
  }

  renderInputFields = () => (
    <Fragment>
      <AddTopic
        disabled={!this.repoSelected}
        relay={this.props.relay}
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
      <AddLink
        disabled={!this.repoSelected}
        relay={this.props.relay}
        topic={this.props.topic}
        viewer={this.props.viewer}
      />
    </Fragment>
  )

  render = () => (
    <form className={this.className}>
      <SelectRepository
        relay={this.props.relay}
        viewer={this.props.viewer}
      />
      { this.repoSelected && this.renderInputFields() }
    </form>
  )
}

export default createFragmentContainer(AddForm, {
  viewer: graphql`
    fragment AddForm_viewer on User {
      selectedRepository {
        isPrivate
      }

      ...AddLink_viewer
      ...AddTopic_viewer
      ...SelectRepository_viewer
    }
  `,
  topic: graphql`
    fragment AddForm_topic on Topic {
      ...AddLink_topic
      ...AddTopic_topic
    }
  `,
})
