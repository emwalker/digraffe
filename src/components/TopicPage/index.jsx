// @flow
import React from 'react'
import { graphql, createFragmentContainer } from 'react-relay'
import { isEmpty } from 'ramda'

import Subhead from 'components/ui/Subhead'
import SidebarList from 'components/ui/SidebarList'
import List from 'components/ui/List'
import type { TopicType } from '../types'
import AddTopic from './AddTopic'
import AddLink from './AddLink'
import { liftNodes } from '../../utils'
import Link from '../ui/Link'
import Topic from '../ui/Topic'

type Props = {
  location: Object,
  orgLogin: string,
  router: Object,
  topic: TopicType,
  viewer: {
    id: string,
  },
}

const TopicPage = ({ location, router, topic, viewer, ...props }: Props) => {
  const {
    childTopics,
    links: childLinks,
    name,
    parentTopics,
  } = topic
  const topics = liftNodes(childTopics)
  const links = liftNodes(childLinks)

  return (
    <div>
      <Subhead
        heading={name}
        location={location}
        router={router}
      />
      <div className="two-thirds column pl-0">
        <List
          placeholder="There are no items in this list."
          hasItems={!isEmpty(topics) || !isEmpty(links)}
        >
          { topics.map(childTopic => (
            <Topic
              key={childTopic.id}
              topic={childTopic}
              {...props}
            />
          )) }

          { links.map(link => (
            <Link
              key={link.id}
              link={link}
              {...props}
            />
          )) }
        </List>
      </div>
      <div className="one-third column pr-0">
        <SidebarList
          title="Parent topics"
          items={liftNodes(parentTopics)}
        />
        <AddTopic
          topic={topic}
          viewer={viewer}
          {...props}
        />
        <AddLink
          topic={topic}
          viewer={viewer}
          {...props}
        />
      </div>
    </div>
  )
}

export const query = graphql`
query TopicPage_query_Query(
  $repoIds: [ID!],
  $topicId: ID!,
  $searchString: String,
) {
  viewer {
    id
  }

  view(repositoryIds: $repoIds) {
    topic(id: $topicId) {
      ...TopicPage_topic @arguments(searchString: $searchString)
    }
  }
}`

export default createFragmentContainer(TopicPage, graphql`
  fragment TopicPage_topic on Topic @argumentDefinitions(
    searchString: {type: "String", defaultValue: ""},
  ) {
    name
    ...AddTopic_topic
    ...AddLink_topic

    parentTopics(first: 100) {
      edges {
        node {
          display: name
          resourcePath
        }
      }
    }

    childTopics(first: 1000, searchString: $searchString) @connection(key: "Topic_childTopics") {
      edges {
        node {
          id
          ...Topic_topic
        }
      }
    }

    links(first: 1000, searchString: $searchString)  @connection(key: "Topic_links") {
      edges {
        node {
          id
          ...Link_link
        }
      }
    }
  }
`)
