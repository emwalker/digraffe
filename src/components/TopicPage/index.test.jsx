import React from 'react'
import { shallow } from 'enzyme'

import TopicPage from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

describe('<TopicPage />', () => {
  const topic = {
    name: 'Frotz',
  }

  const view = {
    repository: {
      displayName: 'Private collection',
      isPrivate: true,
    },
  }

  const wrapper = shallow(
    <TopicPage
      topic={topic}
      view={view}
    />,
  )

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
