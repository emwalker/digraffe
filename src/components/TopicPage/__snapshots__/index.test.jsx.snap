// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopicPage /> renders 1`] = `
<div>
  <Breadcrumbs
    repository={
      Object {
        "displayName": "Private collection",
        "isPrivate": true,
      }
    }
  />
  <Subhead
    heading="Frotz"
  />
  <div
    className="two-thirds column pl-0"
  >
    <Component
      hasItems={false}
      placeholder="There are no items in this list."
    />
  </div>
  <div
    className="one-third column pr-0"
  >
    <Component
      items={Array []}
      title="Parent topics"
    />
    <AddTopic
      topic={
        Object {
          "name": "Frotz",
        }
      }
      view={
        Object {
          "repository": Object {
            "displayName": "Private collection",
            "isPrivate": true,
          },
        }
      }
    />
    <AddLink
      topic={
        Object {
          "name": "Frotz",
        }
      }
      view={
        Object {
          "repository": Object {
            "displayName": "Private collection",
            "isPrivate": true,
          },
        }
      }
    />
  </div>
</div>
`;
